lib/experimental/fonts-library/class-wp-font-family-utils.php

```diff
@@ -0,0 +1,76 @@
+<?php
+/**
+ * Fonts Family Utils class.
+ *
+ * This file contains utils fot Font Family class.
+ *
+ * @package    Gutenberg
+ * @subpackage Fonts Library
+ * @since      X.X.X
+ */
+
+/**
+ * A class of utilities for working with the Font Library.
+ */
+class WP_Font_Family_Utils {
+
+	/**
+	 * Generates a filename for a font face asset.
+	 *
+	 * Creates a filename for a font face asset using font family, style, weight and extension information.
+	 *
+	 * @param array  $font_face The font face array containing 'fontFamily', 'fontStyle', and 'fontWeight' attributes.
+	 * @param string $url The URL of the font face asset, used to derive the file extension.
+	 * @param int    $i Optional counter for appending to the filename, default is 1.
+	 * @return string The generated filename for the font face asset.
+	 */
+	static public function get_filename_from_font_face( $font_face, $url, $i = 1 ) {
+		$extension = pathinfo( $url, PATHINFO_EXTENSION );
+		$family    = sanitize_title( $font_face['fontFamily'] );
+		$style     = sanitize_title( $font_face['fontStyle'] );
+		$weight    = sanitize_title( $font_face['fontWeight'] );
+		$filename  = "{$family}_{$style}_{$weight}";
+		if ( $i > 1 ) {
+			$filename .= "_{$i}";
+		}
```

Comment:
I assume the `$i` is intended to provide a unique filename, right? If yes, Core has a native function to handle this: `wp_unique_filename()`. This Core function also sanitizes the filename too.

Reply:
The `$i` is to add a number if a font face has more than one asset linked. I think `wp_unique_filename` would not work for this because the intention here is to overwrite the font asset if it already exists. For example, if the user installs the same font face twice, let's say Piazzolla 500 Italic, we want to overwrite the asset and have only one `piazzolla_500_italic.ttf` file instead of 2.

Reply:
> Do you mean if it has multiple assets like a remote and a local version?

No, precisely, but yes, a font face can have multiple assets.

> If not, can you elaborate on the use case

`fontFace.src` can be a string (only one path/URL) or an array (multiple URLs or paths). The case when a font face defines two assets with the same file extension is rare but can be possible, and due to the fact we are naming the assets based on the font face definition, we need to add the index to avoid writing the same file twice.

If our `theme.json` data looks like this:
```
{
    "fontFamily" => "Piazzolla",
    "fontStyle" => "italic",
    "fontWeight" => "400",
    "src" => [
        "https://example.com/fonts/piazzolla.ttf",
        "https://example.com/fonts/piazzolla.woff2"
    ]
}
```

We print this kind of CSS to the page:
```
@font-face {
     font-family: "Piazzolla";
    src: url("https://example.com/fonts/piazzolla.ttf") format("ttf"), 
        url("https://example.com/fonts/piazzolla.woff2") format("woff2");
    font-weight: 400;
    font-style: italic;
}
```  


> We try not to use single character variable names, especially for parameters, and have made an effort over the past few releases to replace these in Core.

Good, we can rename it.

Reply:
Thanks @matiasbenedetto!

This seems less about filesystems and more about unique normalization then. I wonder, if the additional data for the name must be passed by the developer, would `(string) $suffix` be an appropriate name? That way, a developer could choose whether they want to pass `2...3...4`, or something like `parse_url( $url )['host']`, for example.

In terms of sanitization:
- If the integer type stays, then there should be a type check with a `trigger_error()`/`_doing_it_wrong()`, as it currently accepts any value that evaluates to `> 1`, which is effectively any non-empty, non-numeric string.
- If the type changes to string, the filename needs to be sanitized before being returned.

Why?
- Not type checking locks us into the BC promise for non-integer types (string, in this case).
- Not sanitizing the filename produces a possible security risk, as the _exact_ appended data may not be entirely under the total control of the developer, or may include characters that are unacceptable in some operating systems.
- I'd suggest removing the `sanitize_title()`, and returning `sanitize_file_name( "{$filename}.{$extension}" );`.
  - `sanitize_file_name()` calls `sanitize_title_with_dashes()` internally.
  - `sanitize_file_name()` handles special characters that should be removed for cross-compatibility.

Reply:
Great suggestions. Done!